File: app/src/main/java/com/dragonpav/dengine/Object3D.java
package com.dragonpav.dengine;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.opengl.GLES30;
import android.util.Log;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;

import glm_.mat4x4.Mat4;

public class Object3D {
	public Values.Vector3 position;
	public Values.Matrix4 model;
	private Camera cam;
	private RenderUtils renderUtils;
	public FloatBuffer vertexBuffer = null;
	public IntBuffer indexBuffer = null;
	public FloatBuffer texBuffer = null;
	public FloatBuffer normalBuffer = null;
	public int vertexBufferSize, indexBufferSize, texBufferSize, normalBufferSize;
	public Texture texture;
	public ObjectBuffer b;
	public BoundingBox bounds;
	public int polygonMode = GLES30.GL_TRIANGLES;
	public void begin() {
		if (renderUtils.lighting != null) {
			renderUtils.lighting.bind();
		}
		model.values = new Mat4(1.0f);
		GLES30.glBindVertexArray(b.VAO);
	}
	public void render() {
		model.uniform();
		GLES30.glActiveTexture(texture.glType);
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, texture.textureId);
		if (cam.autoUpdate) {
			cam.update();
		}
		if (texture.texSampleNum != -1) {
			GLES30.glUniform1i(texture.texSampleNum, texture.glType % GLES30.GL_TEXTURE0);
		}
		GLES30.glDrawElements(polygonMode, indexBufferSize / 4, GLES30.GL_UNSIGNED_INT, 0);
		int err;
		while ((err = GLES30.glGetError()) != GLES30.GL_NO_ERROR) {
			Log.e("DEngine", String.format("OpenGL error occured: 0x%X", err));
		}
	}
	public void end() {
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, 0);
		GLES30.glBindVertexArray(0);
	}
	public void setup() {
		b = renderUtils.createStaticVAO(vertexBuffer, indexBuffer, texBuffer, normalBuffer,
				vertexBufferSize, indexBufferSize, texBufferSize, normalBufferSize);
	}
	public void dispose() {
		end();
		if (b.VAO != 0) {
			int[] vaoArray = new int[] {b.VAO};
			GLES30.glDeleteVertexArrays(1, vaoArray, 0);
			b.VAO = 0;
		}
		int count = 0;
		if (b.VBO != 0) count++;
		if (b.VBOtexture != 0) count++;
		if (b.VBOnormal != 0) count++;
		if (b.EBO != 0) count++;
		if (count > 0) {
			int[] buffersToDelete = new int[count];
			int index = 0;
			if (b.VBO != 0) {
				buffersToDelete[index++] = b.VBO;
			}
			if (b.VBOtexture != 0) {
				buffersToDelete[index++] = b.VBOtexture;
			}
			if (b.VBOnormal != 0) {
				buffersToDelete[index++] = b.VBOnormal;
			}
			if (b.EBO != 0) {
				buffersToDelete[index++] = b.EBO;
			}
			b = new ObjectBuffer();
			GLES30.glDeleteBuffers(count, buffersToDelete, 0);
		}
	}
	public Object3D(Camera camera, RenderUtils rend, Texture tex) {
		cam = camera;
		renderUtils = rend;
		texture = tex;
		model = new Values.Matrix4(GLES30.glGetUniformLocation(rend.p.programId, "model"));
	}
}

File: src/Object3D.cpp
#include "Object3D.h"
Object3D::Object3D(Camera* camera, RenderUtils* rend, Texture* tex) {
	if (camera == nullptr) {
		throw std::runtime_error("Camera is null");
	}
	if (rend == nullptr) {
		throw std::runtime_error("RenderUtils is null");
	}
	if (tex == nullptr) {
		throw std::runtime_error("Texture is null");
	}
	this->cam = camera;
	renderUtils = rend;
	texture = tex;
	model.values = glm::translate(glm::mat4(1.0f), glm::vec3(position.getX(), position.getY(), position.getZ()));
	modelLocation = glGetUniformLocation(rend->getProgram()->getId(), "model");
}
Object3D::~Object3D() {
	dispose();
}
void Object3D::setPosition(Vector3 pos) {
	position = pos;
}
void Object3D::setBounds(BoundingBox bb) {
	bounds = bb;
}
void Object3D::setTexture(Texture* tex) {
	texture = tex;
}
void Object3D::setCamera(Camera* camera) {
	cam = camera;
}
void Object3D::setPolygonMode(GLenum mode) {
	polygonMode = mode;
}
Vector3& Object3D::getPosition() {
	return position;
}
BoundingBox& Object3D::getBounds() {
	return bounds;
}
Texture* Object3D::getTexture() {
	return texture;
}
Camera* Object3D::getCamera() {
	return cam;
}
GLenum Object3D::getPolygonMode() {
	return polygonMode;
}
Matrix4& Object3D::getModelMatrix() {
	return model;
}
void Object3D::setup() {
	GLuint verticesSize = (GLuint) vertexBuffer.size() * sizeof(GLfloat);
	GLuint texCoordsSize = (GLuint) texBuffer.size() * sizeof(GLfloat);
	GLuint indicesSize = (GLuint) indexBuffer.size() * sizeof(GLuint);
	GLuint normalsSize = (GLuint) normalBuffer.size() * sizeof(GLfloat);
	b = renderUtils->createStaticVAO(vertexBuffer.data(), indexBuffer.data(), texBuffer.data(), normalBuffer.data(), verticesSize, indicesSize, texCoordsSize, normalsSize);
}
void Object3D::begin() {
	glBindVertexArray(b.VAO);
}
void Object3D::render(Program* currentProgram, bool isShadowPass) {
	if (currentProgram == nullptr) return;

	if (!isShadowPass) {
		if (!cam->isBoxVisible(bounds)) {
			return;
		}
	}

	if (isShadowPass) {
		if (shadowModelLocation == -1) {
			shadowModelLocation = glGetUniformLocation(currentProgram->getId(), "model");
		}
		model.uniformL = shadowModelLocation;
		model.uniform();
	} else {
		model.uniformL = modelLocation;
		model.uniform();
	}

	if (isShadowPass) {
		glDrawElements(polygonMode, (GLuint)indexBuffer.size(), GL_UNSIGNED_INT, 0);
		return;
	}

	if (normalMatrixLocation == -1) {
		normalMatrixLocation = glGetUniformLocation(currentProgram->getId(), "normalMatrix");
	}

	if (normalMatrixLocation != -1) {
		glm::mat3 normalMatrix = glm::transpose(glm::inverse(glm::mat3(model.values)));
		glUniformMatrix3fv(normalMatrixLocation, 1, GL_FALSE, glm::value_ptr(normalMatrix));
	}

	glActiveTexture(texture->getGLType());
	glBindTexture(GL_TEXTURE_2D, texture->getTextureId());
	if (this->cam->isAutoUpdate()) {
		this->cam->update();
	}
	if (cam->getRenderUtils()->getLighting() != nullptr) {
		cam->getRenderUtils()->getLighting()->bind();
	}
	if (texture->getTextureSample() == -1) {
		std::cerr << "texture.texSampleNum is -1" << std::endl;
		return;
	}
	glUniform1i(texture->getTextureSample(), texture->getGLType() % GL_TEXTURE0);
	glDrawElements(polygonMode, (GLuint) indexBuffer.size(), GL_UNSIGNED_INT, 0);
	GLenum err;
	while ((err = glGetError()) != GL_NO_ERROR) {
		std::cout << "OpenGL error occured: 0x" << std::hex << err << std::dec << std::endl;
	}
	glBindVertexArray(0);
	glBindTexture(GL_TEXTURE_2D, 0);
}
void Object3D::dispose() {
	glDeleteVertexArrays(1, &b.VAO);
	std::vector<GLuint> buffersToDelete;
	buffersToDelete.push_back(b.VBO);
	buffersToDelete.push_back(b.VBOtexture);
	buffersToDelete.push_back(b.EBO);
	if (cam->getRenderUtils()->getLighting() != nullptr) buffersToDelete.push_back(b.VBOnormal);
	glDeleteBuffers((GLuint) buffersToDelete.size(), buffersToDelete.data());
	b = ObjectBuffer();
}